### PR TITLE
A Service Operator can remove problematic payments from a payroll run

### PR DESCRIPTION
**Pull request by pezholio**
This allows service operators to remove payments from a Payroll Run if Cantium have confirmed there is some problem (i.e. incorrect bank details) that has stopped a payment from taking place. I've also updated the guidance that a user sees if a payment has been missed off and the confirmation report cannot be uploaded.

Once a confirmation report has been uploaded, it's no longer possible to remove payments from a run

# Screenshots

## Before a payroll run has been confirmed

image

## After a payroll run has been confirmed

image

## Before deleting a payment

image

## After a payment has been deleted 

image

### CODE REVIEW

**Comment by pezholio** [comment]
[A service operator can upload a payment confirmation report that is missing records](https://trello.com/c/WrFM4sa3/1253-a-service-operator-can-upload-a-payment-confirmation-report-that-is-missing-records)

File: CHANGELOG.md [hunk 1]
```diff
@@ -31,6 +31,7 @@ The format is based on [Keep a Changelog]
 - Emails are sent for unchecked claims three weeks after they are submitted,
   advising users that their claim is still in progress
 - Developers can export data needed for school check mail merge
+- A Service Operator can remove problematic payments from a payroll run
```
Comment (tekin) [inline CHANGELOG.md]: This is not in the `Unreleased` section.

File: app/assets/stylesheets/components/inline-form.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+.inline-form {
```
Comment (tekin) [inline app/assets/stylesheets/components/inline-form.scss]: I reckon we can do this without the need for an additional SCSS file and class. Worth speaking to @titlescreen about how to do that?

File: app/controllers/admin/payments_controller.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  class PaymentsController < BaseAdminController
+    before_action :ensure_service_operator
+    before_action :find_payroll_run
+    before_action :find_payment
+
+    def delete
+    end
+
+    def destroy
+      if @payroll_run.confirmation_report_uploaded_by.nil?
```
Comment (tekin) [inline app/controllers/admin/payments_controller.rb]: Is there an undefined concept/idea here that we could encapsulate in a method on `PayrollRun` to better communicate what is going on here?

File: app/views/admin/payments/destroy.html.erb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        <span class="govuk-warning-text__assistive">Warning</span>
+          <%= t("admin.payment_deleted_claim_next_steps",
+              count: @claim_references.count,
+              references: @claim_references.to_sentence) %>
+        </span>
+      </strong>
+    </div>
+
+    <p class="govuk-body">
+      <%= link_to "Back to payroll run", admin_payroll_run_path(@payroll_run) %>
```
Comment (tekin) [inline app/views/admin/payments/destroy.html.erb]: This link needs a class.

File: app/models/payment.rb [hunk 5]
```diff
@@ -1,5 +1,5 @@
 class Payment < ApplicationRecord
-  has_many :claims
+  has_many :claims, dependent: :nullify
```
Comment (tekin) [inline app/models/payment.rb]: Why was this removed?

File: app/helpers/admin/payment_helper.rb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+module Admin
+  module PaymentHelper
+    def list_references(payment)
```
Comment (tekin) [inline app/helpers/admin/payment_helper.rb]:
```suggestion
    def claim_references(payment)
```

File: app/views/admin/payments/delete.html.erb [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column-full">
+    <h1 class="govuk-heading-l">
+    Are you sure you want to remove this payment from the payroll run?
```
Comment (tekin) [inline app/views/admin/payments/delete.html.erb]:
```suggestion
      Are you sure you want to remove this payment from the payroll run?
```

File: app/views/admin/payments/delete.html.erb [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    </h1>
+
+    <p class="govuk-body-l">
+      <%= t("admin.payment_deletion_claim_detail",
```
Comment (tekin) [inline app/views/admin/payments/delete.html.erb]: nice use of locals for this, I always forget about the one/other stuff 👍🏽

File: config/locales/en.yml [hunk 9]
```diff
@@ -78,6 +78,15 @@ en:
     claims_preventing_payment_message:
       one: This claim cannot currently be approved because we’re already paying another claim (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
       other: This claim cannot currently be approved because we’re already paying other claims (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
+    payment_deletion_claim_detail:
```
Comment (tekin) [inline config/locales/en.yml]: How about a more descriptive name for this?


```suggestion
    payment_deletion_claim_warning:
```

File: config/locales/en.yml [hunk 10]
```diff
@@ -78,6 +78,15 @@ en:
...
+      one: This will remove the claim %{references} from the payroll run. Make a note of this claim so that it can be corrected for the next payroll run.
+      other: This will remove the claims %{references} from the payroll run. Make a note of these claims so that they can be corrected for the next payroll run.
+    payment_deleted_claim_detail:
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
    payment_deleted_outcome_message:
```

File: config/locales/en.yml [hunk 11]
```diff
@@ -78,6 +78,15 @@ en:
...
+      one: Claim reference %{references} has been removed from this payroll run.
+      other: Claim references %{references} have been removed from this payroll run.
+    payment_deleted_claim_next_steps:
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
    payment_deleted_next_steps_message:
```

File: app/views/admin/payroll_runs/show.html.erb [hunk 12]
```diff
@@ -64,4 +64,40 @@
...
+      <% @payroll_run.payments.each do |payment| %>
+        <tr class="govuk-table__row">
+          <th scope="row" class="govuk-table__header"><%= payment.id %></th>
+          <td class="govuk-table__cell"><%= list_references(payment) %></td>
+          <td class="govuk-table__cell"><%= payment.banking_name %></td>
+          <% unless @payroll_run.confirmation_report_uploaded_by %>
+            <td class="govuk-table__cell">
+              <%= link_to 'Remove',
+                          delete_admin_payroll_run_payment_path(id: payment.id, payroll_run_id: payment.payroll_run.id),
+                          class: "govuk-button govuk-button--warning"
```
Comment (tekin) [inline app/views/admin/payroll_runs/show.html.erb]: Should these be styled as links, not buttons?

File: app/controllers/admin/payments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,31 @@
+module Admin
+  class PaymentsController < BaseAdminController
+    before_action :ensure_service_operator
+    before_action :find_payroll_run
+    before_action :find_payment
+
+    def delete
```
Comment (tekin) [inline app/controllers/admin/payments_controller.rb]: We've used the word "remove" in the UI. Should we name this action that too so it matches and makes it all link up?

File: config/routes.rb [hunk 14]
```diff
@@ -95,6 +95,9 @@ def matches?(request)
     resources :payroll_runs, only: [:index, :new, :create, :show] do
       resources :payment_confirmation_report_uploads, only: [:new, :create]
       resource :download, only: [:new, :create, :show], controller: "payroll_run_downloads"
+      resources :payments, only: [:destroy] do
+        get :delete, to: "payments#delete", on: :member
```
Comment (tekin) [inline config/routes.rb]: We can rely on the router conventions to map to the controller and action without being explicit:

```suggestion
        get :delete, on: :member
```

File: app/controllers/admin/payments_controller.rb (hunk 3 above)
Comment (pezholio) [reply app/controllers/admin/payments_controller.rb]: 👍 I've added a method to `PayrollRun`:

```
def confirmation_report_uploaded?
  confirmation_report_uploaded_by.present?
end
```

File: app/models/payment.rb (hunk 5 above)
Comment (pezholio) [reply app/models/payment.rb]: I'm not sure what you mean here? I've had to add `dependent: :nullify` because deleting payments results in an `ActiveRecord::InvalidForeignKey` error, because the Payment ID isn't removed from the claim on destroy.

File: app/views/admin/payments/delete.html.erb (hunk 8 above)
Comment (pezholio) [reply app/views/admin/payments/delete.html.erb]: It's super useful 👍

File: app/views/admin/payroll_runs/show.html.erb (hunk 12 above)
Comment (pezholio) [reply app/views/admin/payroll_runs/show.html.erb]: Yeah, that makes sense, given we link to a confirmation page now

File: app/controllers/admin/payments_controller.rb (hunk 13 above)
Comment (pezholio) [reply app/controllers/admin/payments_controller.rb]: That's much neater, ta 👍

File: app/models/payment.rb (hunk 5 above)
Comment (tekin) [reply app/models/payment.rb]: Oh, I mistyped, I meant added. I was wondering why it was added as I was under the impression that the default was `:nullify` for `has_many` relations but it looks like I'm wrong about that!

File: app/models/payroll_run.rb [hunk 15]
```diff
@@ -34,6 +34,10 @@ def download_available?
     download_triggered? && Time.zone.now - downloaded_at < DOWNLOAD_FILE_TIMEOUT.seconds
   end
 
+  def confirmation_report_uploaded?
```
Comment (tekin) [inline app/models/payroll_run.rb]: Now we have this method I reckon there are a couple of other places we could be using it (see `payroll_runs/index` and `payroll_runs/show`).

File: spec/requests/admin_payroll_run_payments_spec.rb [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+  let(:admin) { create(:dfe_signin_user) }
+
+  before do
+    sign_in_to_admin_with_role(DfeSignIn::User::SERVICE_OPERATOR_DFE_SIGN_IN_ROLE_CODE, admin.dfe_sign_in_id)
+  end
+
+  let(:payroll_run) { create(:payroll_run, claims_counts: {MathsAndPhysics => 1, StudentLoans => 1}) }
+  let(:payment) { payroll_run.payments.first }
+
+  describe "delete" do
```
Comment (tekin) [inline spec/requests/admin_payroll_run_payments_spec.rb]:
```suggestion
  describe "remove" do
```

File: spec/requests/admin_payroll_run_payments_spec.rb [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+    it "shows a confirmation screen" do
```
Comment (tekin) [inline spec/requests/admin_payroll_run_payments_spec.rb]: This spec doesn't actually have any assertions in it.

File: spec/requests/admin_payroll_run_payments_spec.rb [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+      get remove_admin_payroll_run_payment_path(
+        payroll_run_id: payroll_run.id,
+        id: payment.id
+      )
+    end
+  end
+
+  describe "destroy" do
+    it "deletes a payroll run and redirects with a message" do
```
Comment (tekin) [inline spec/requests/admin_payroll_run_payments_spec.rb]: Does it redirect?

File: spec/requests/admin_payroll_run_payments_spec.rb [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+      expect {
+        delete admin_payroll_run_payment_path(
+          payroll_run_id: payroll_run.id,
+          id: payment.id
+        )
+      }.to change(payroll_run.reload.payments, :count).by(-1)
+
+      expect(payroll_run.payments).to_not include(payment)
+
+      expect(response.body).to include("Payment has been removed from payroll run")
```
Comment (tekin) [inline spec/requests/admin_payroll_run_payments_spec.rb]: I think this spec and the feature spec should confirm the the claim references are displayed as it's an important part of the feature.

File: app/models/payment.rb (hunk 5 above)
Comment (pezholio) [reply app/models/payment.rb]: 👍

File: spec/features/admin_payroll_runs_spec.rb [hunk 20]
```diff
@@ -91,6 +91,45 @@
...
+
+    click_on "Remove", match: :first
+
+    expect(page).to have_content("Are you sure you want to remove payment #{payroll_run.payments.first.id} from the payroll confirmation report?")
+
+    expect {
+      click_on "Remove payment"
+    }.to change(payroll_run.reload.payments, :count).by(-1)
+
+    expect(page).to have_content("Payment has been removed from payroll run")
```
Comment (tekin) [inline spec/features/admin_payroll_runs_spec.rb]: We should also confirm the specific payment was removed, rather than just that the total count was decreased by one, plus that the specific claim references appear on the page.

File: spec/features/admin_payroll_runs_spec.rb [hunk 21]
```diff
@@ -91,6 +91,45 @@
...
     expect(page).to have_field("payroll_run_download_link", with: new_admin_payroll_run_download_url(payroll_run))
+
+    expect(page).to have_content payroll_run.payments[0].id
+    expect(page).to have_content payroll_run.payments[1].id
+
+    expect(page).to have_link(href: remove_admin_payroll_run_payment_path(id: payroll_run.payments[0].id, payroll_run_id: payroll_run.id))
+    expect(page).to have_link(href: remove_admin_payroll_run_payment_path(id: payroll_run.payments[1].id, payroll_run_id: payroll_run.id))
+  end
+
+  scenario "Service operator cannot see removal buttons when confirmation report has been uploaded" do
```
Comment (tekin) [inline spec/features/admin_payroll_runs_spec.rb]: I think this scenario is probably redundant because the ` "Service operator can remove a payment from a payroll run"` wouldn't work unless those remove links are on the page.

Comment (pezholio) [reply spec/features/admin_payroll_runs_spec.rb]: This is testing the reverse now. That said, there are links, not buttons now, and the spec still passes, so maybe it's not worth having now 😆

File: config/locales/en.yml [hunk 22]
```diff
@@ -78,6 +78,18 @@ en:
     claims_preventing_payment_message:
       one: This claim cannot currently be approved because we’re already paying another claim (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
       other: This claim cannot currently be approved because we’re already paying other claims (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
+    payment_deletion_claim_warning:
+      one: This means claim %{references} has not been included in this pay run.
+      other: This means claims %{references} have not been included in this pay run.
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      other: This means claims %{references} will not be included in this pay run.
```

File: config/locales/en.yml [hunk 23]
```diff
@@ -78,6 +78,18 @@ en:
     claims_preventing_payment_message:
       one: This claim cannot currently be approved because we’re already paying another claim (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
       other: This claim cannot currently be approved because we’re already paying other claims (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
+    payment_deletion_claim_warning:
+      one: This means claim %{references} has not been included in this pay run.
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      one: This means claim %{references} will not be included in this pay run.
```

File: config/locales/en.yml [hunk 24]
```diff
@@ -78,6 +78,18 @@ en:
...
+      other: This means claims %{references} have not been included in this pay run.
+    payment_deleted_title:
+      one: You have removed a claim from the payroll confirmation report
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      one: You have removed a claim from the payroll run
```

File: config/locales/en.yml [hunk 25]
```diff
@@ -78,6 +78,18 @@ en:
...
+      other: You have removed %{count} claims from the payroll confirmation report
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      other: You have removed %{count} claims from the payroll run
```

File: config/locales/en.yml [hunk 26]
```diff
@@ -78,6 +78,18 @@ en:
...
+    payment_deleted_outcome_message:
+      one: Claim %{references} has not been paid in this pay run.
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      one: Claim %{references} will not be paid in this pay run.
```

File: config/locales/en.yml [hunk 27]
```diff
@@ -78,6 +78,18 @@ en:
...
+      other: Claims %{references} have not been paid in this pay run.
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      other: Claims %{references} will not be paid in this pay run.
```

File: app/views/admin/payroll_runs/show.html.erb [hunk 28]
```diff
@@ -64,4 +64,40 @@
...
+  <div class="govuk-grid-column-full">
+
+    <h2 class="govuk-heading-l">Payments</h2>
+
+    <table class="govuk-table">
+    <thead class="govuk-table__head">
+      <tr class="govuk-table__row">
+        <th scope="col" class="govuk-table__header">Payment ID</th>
+        <th scope="col" class="govuk-table__header">Claim Reference(s)</th>
+        <th scope="col" class="govuk-table__header">Applicant Name</th>
```
[the remark this answers is not in the document]
Comment (pezholio) [reply app/views/admin/payroll_runs/show.html.erb]: I'm gonna go with `Payee Name` here as I think that expresses it a bit better

File: config/locales/en.yml (hunk 22 above)
Comment (pezholio) [reply config/locales/en.yml]: Yeah, I talked to Faye about this, I think her thinking was that we are deleting this because it hasn't been included in the confirmation report, but I agree, it is a bit confusing.

File: config/locales/en.yml (hunk 26 above)
Comment (pezholio) [reply config/locales/en.yml]: I get where you're coming from here, the claim not being paid is not a side effect of us deleting the payment. This has already been decided by Cantium

**Comment by tekin** [comment]
Faye's given us the thumbs up to tweak the content as we feel 👍🏽

File: config/locales/en.yml [hunk 29]
```diff
@@ -78,6 +78,18 @@ en:
...
       one: This claim cannot currently be approved because we’re already paying another claim (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
       other: This claim cannot currently be approved because we’re already paying other claims (%{claim_references}) to this claimant in this payroll month using different payment details. Please speak to a Grade 7.
+    payment_deletion_claim_warning:
+      one: This means claim %{references} will no longer be included in this pay run.
+      other: This means claims %{references} will no longer be included in this pay run.
+    payment_deleted_title:
+      one: You have removed a claim from the payroll run
+      other: You have removed %{count} claims from the payroll run
+    payment_deleted_outcome_message:
+      one: Claim %{references} are no longer included in this pay run.
```
Comment (tekin) [inline config/locales/en.yml]:
```suggestion
      one: Claim %{references} is no longer included in this pay run.
```

Comment (pezholio) [reply config/locales/en.yml]: Gaaaaah
